### ISSUE TITLE
shutdown should accept a date, rather than just a time

### ISSUE THREAD

**Issue by barak** [systemd/systemd#11928]
This is a small feature suggestion, basically for symmetry.

Thanks to the systemd and other robust mechanisms, my Debian computers usually run without rebooting for months and sometimes even years.

But sometimes some external issue, like a scheduled power outage, mandates a clean scheduled shutdown. This just happened to me: there is a scheduled power outage at my workplace at 04:00 about two weeks from today, so I need to power down a handful of machines beforehand.

Naturally I’d wish to
````
$ sudo shutdown --poweroff '2019-03-19 03:00'
````
But that doesn’t work, shutdown only accepts a time, or a delay in minutes. Instead I end up converting the delay to minutes:
````
$ sudo shutdown --poweroff +$(echo "($(date --date='2019-03-19 03:00' '+%s') - $(date '+%s')) / 60" | bc)
````
Yuck. The shutdown executable (which is just a symlink to systemctl, so in fact already links to date parsing routines) really should take a proper date string.

Perhaps this functionality can be accessed with `systemctl poweroff --some-option=TIMESPEC`, although I didn’t see it in `systemctl(1)`. It might make sense to add that, either the option or the documentation. It might also make sense for `shutdown(1)` to mention `systemctl shutdown`.

**Comment by poettering** [comment]
I am sorry, but I am very sure that we shouldn't extend the shutdown compat tool. It's provided for compat only, it has an interface not defined by us, but by sysvinit, we hence should not extend it.

what's wrong with doing this:

```
systemd-run --on-calendar="2019-03-19 03:00" systemctl poweroff
```

That appears a lot more generic and powerful, isn't that entirely sufficient for what you are trying to do?

**Comment by doverride** [comment]
That wouldnt take pam_nologin into account. I suppose something like `echo "Poweroff scheduled for 2019-03-19 03:00, no more logins allowed." > /run/nologin && systemd-run --on-calendar="2019-03-19 03:00" systemctl poweroff`

**Comment by mbiebl** [comment]
@poettering Could this be added to `systemctl shutdown` but not `shutdown`. Something like `systemctl shutdown --when=xxxx`

**Comment by barak** [comment]
@poettering I certainly understand the motivation for not extending the API of `shutdown`. But it would be nice to clearly document the non-compat systemd-appropriate way to achieve this objective in `shutdown(1)`, as a way to transition people like me off the compat tools.

I think that `shutdown TIME` differs from `systemd-run --on-calendar=TIME systemctl poweroff` in that the former will not survive a power cycle, whereas the latter will. Also `shutdown` blats periodic shutdown warnings on all terminals 1970s-style, which although totally frivolous is something I do rather enjoy.
````
Broadcast message from root@tarsk on pts/10 (Sun 2019-03-10 03:01:40 GMT):

The system is going down for poweroff at Tue 2019-03-19 03:01:40 GMT!


Broadcast message from root@tarsk on pts/10 (Sun 2019-03-10 04:01:40 GMT):

The system is going down for poweroff at Tue 2019-03-19 03:01:40 GMT!


Broadcast message from root@tarsk on pts/10 (Sun 2019-03-10 05:01:40 GMT):

The system is going down for poweroff at Tue 2019-03-19 03:01:40 GMT!


Broadcast message from root@tarsk on pts/10 (Sun 2019-03-10 06:01:40 GMT):

The system is going down for poweroff at Tue 2019-03-19 03:01:40 GMT!


Broadcast message from root@tarsk on pts/10 (Sun 2019-03-10 07:01:40 GMT):

The system is going down for poweroff at Tue 2019-03-19 03:01:40 GMT!


Broadcast message from root@tarsk on pts/10 (Sun 2019-03-10 08:01:40 GMT):

The system is going down for poweroff at Tue 2019-03-19 03:01:40 GMT!


Broadcast message from root@tarsk on pts/10 (Sun 2019-03-10 09:01:40 GMT):

The system is going down for poweroff at Tue 2019-03-19 03:01:40 GMT!

$ date
Sun Mar 10 09:14:13 GMT 2019
$ 
````

**Comment by boucman** [comment]
systemd-run created timers survive a power cycle ? I thought they were created in /run

**Comment by barak** [comment]
I thought so, but on rereading I don't think they even survive the invoking user logging out, unless lingering is enabled. Even then, maybe `systemctl` will want a password, so this wouldn't automate the process. Anyway, I have no solid idea. Is this something that is apparent in the documentation? I'm having trouble figuring it out from `systemd-run(1)` which is where such an important bit of information would, I'd imagine, be made clear.

**Comment by boucman** [comment]
in the documentation, it is said that the units created are transient, so they shouldn't survive reboot.

I am a bit suprised they disapear on logout or that linger makes a diffrence. That would make sense if you run systemd-run --user, since that explicitely would tie them to the user session. Normal systemd-run invocations would be linked to the system instance and would not be linked to the session lifecycle.

As for asking for password, i'm not sure how you did that with cron in the first place, but there is the systemd-ask-password for that sort of cases.

**Comment by poettering** [comment]
> @poettering Could this be added to `systemctl shutdown` but not `shutdown`. Something like `systemctl shutdown --when=xxxx`

but why? what makes shutdown so special? i mean, we wouldn't add "systemctl start foo.service --when=xxxx" either, would we?

**Comment by poettering** [comment]
> I think that `shutdown TIME` differs from `systemd-run --on-calendar=TIME systemctl poweroff` in that the former will not survive a power cycle, whereas the latter will.

I cannot parse this. How can a request to shut down survive a power cycle? When the shutdown happens this is the end of the power cycle, by definition, so of course it won't survive it? I really don't grok this sentence...

**Comment by poettering** [comment]
> I thought so, but on rereading I don't think they even survive the invoking user logging out, unless lingering is enabled.

"systemd-run" without "--user" creates a system timer, and that is independent of any user logging out or in. If you pass "--user" what yo are saying would be true, but this is not what the cmdline I pasted does, no?

**Comment by mbiebl** [comment]
> but why? what makes shutdown so special? i mean, we wouldn't add "systemctl start foo.service --when=xxxx" either, would we?

The main point is convenience, I guess, and the support for wall messages.
Both you don't get via `systemd-run`
You raise an interesting point though. I could imagine that `systemctl start --when` might actually be useful

**Comment by barak** [comment]
More elaborate timing facilities would certainly by useful for things like https://github.com/kdave/btrfsmaintenance which currently uses a Rube-Goldberg set of hacks to schedule clean-up jobs periodically, and even so doesn't bother to prevent undesirable concurrency because that would require yet more brittle hacks. So if you do add such general scheduling facilities, I'd suggest checking a couple more use cases, aside from just scheduled shutdowns, to make sure they're all well handled.

**Comment by poettering** [comment]
We have timer units for that already, and systemd-run --on-calendar= makes these things easily accessible. I am not convinced this usecase is strong enough to add any more explicit support for anything like this. I mean, I think it's fine making common stuff easier by adding some shortcut glue, but this does not strike me a case like this, and I'd rather have fewer magic hookups here. There's nothing wrong with using "systemd-run --on-calendar=" and it's so very generic.

**Comment by barak** [comment]
Fair enough. Consider this just a documentation suggestion, then.

**Comment by poettering** [comment]
I guess fixed by: 1433e1f998465b7acf472c73d58c14e7e2eb3f13
